cell2location/distributions/AutoNormal.py

```diff
@@ -0,0 +1,312 @@
+        result = {}
+
+        # sample sites without hierarchical dependency
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            # Don't sample if the site has hierarchical dependency
+            if name in self.hierarchical_sites.keys():
+                continue
+            transform = biject_to(site["fn"].support)
+
+            with ExitStack() as stack:
+                for frame in site["cond_indep_stack"]:
+                    if frame.vectorized:
+                        stack.enter_context(plates[frame.name])
+
+                site_loc, site_scale = self._get_loc_and_scale(name)
+                unconstrained_latent = pyro.sample(
+                    name + "_unconstrained",
+                    dist.Normal(
+                        site_loc,
+                        site_scale,
+                    ).to_event(self._event_dims[name]),
+                    infer={"is_auxiliary": True},
+                )
+
+                value = transform(unconstrained_latent)
+                if poutine.get_mask() is False:
+                    log_density = 0.0
+                else:
+                    log_density = transform.inv.log_abs_det_jacobian(
+                        value,
+                        unconstrained_latent,
+                    )
+                    log_density = sum_rightmost(
+                        log_density,
+                        log_density.dim() - value.dim() + site["fn"].event_dim,
+                    )
+                delta_dist = dist.Delta(
+                    value,
+                    log_density=log_density,
+                    event_dim=site["fn"].event_dim,
+                )
+
+                result[name] = pyro.sample(name, delta_dist)
+
+        # Sample sites with hierarchical dependency
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            # Don't sample if the site does not have a hierarchical dependency
+            if name not in self.hierarchical_sites.keys():
+                continue
+            transform = biject_to(site["fn"].support)
+
+            # Get the expected value of the site based on hierarchy
+            # Get values of parent sites
+            parent_names = self.hierarchical_sites[name]
+            parent_result = {k: result[k] for k in parent_names}
+
+            # Propagate through a section of the model (block)
+            # to get the expected value of the site
+            with poutine.block():
+                model_block = poutine.block(self.model, expose=[name] + parent_names)
+                conditioned = poutine.condition(model_block, data=parent_result)
+                conditioned_trace = poutine.trace(conditioned).get_trace(*args, **kwargs)
+                site_loc_hierarhical_constrained = conditioned_trace.nodes[name]["value"]
+
+            # transform to unconstrained space
+            site_loc_hierarhical_unconstrained = transform.inv(site_loc_hierarhical_constrained)
+
```

Comment:
Hi @fritzo

Would be fantastic to get your feedback on the correctness of this procedure for automatically constructing hierarchy in the guide.